### PR TITLE
fix(clients/java): make a `OAuthCredentialsProvider` class thread-safe

### LINKED ISSUE #11885: `io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache` is not thread safe.
**Describe the bug**

After https://github.com/camunda/zeebe/issues/11816 I never thought that the `io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider` is not a thread-safe. So, when you start multiple `JobWorker` in parallel with `OAuthCredentialsProvider` you got a problem in this method: https://github.com/camunda/zeebe/blob/28fe054b79fcc7398a086a1576199efef01a122c/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java#L87-L111

So, in there is the first checking that the file exists, and then there is a creating one if none exists. If this part of the code is executed concurrently we get an error that we are trying to create a file that already exists.

**To Reproduce**

1. Get a fresh main branch.
2. Configure Zeebe Client with `OAuthCredentialsProvider`
3. Start some requests in the parallel

**Expected behavior**

If requests with the Zeebe Client and `OAuthCredentialsProvider` runned at parallel there is no error about a file, that already created.  

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
io.grpc.StatusException: CANCELLED
	at io.grpc.Status.asException(Status.java:554)
	at io.grpc.kotlin.ClientCalls$rpcImpl$1$1$1.onClose(ClientCalls.kt:296)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:563)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:744)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.nio.file.FileAlreadyExistsException: /root/.camunda/credentials
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:94)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:218)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:375)
	at java.base/java.nio.file.Files.createFile(Files.java:652)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache.ensureCacheFileExists(OAuthCredentialsCache.java:110)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache.writeCache(OAuthCredentialsCache.java:69)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.refreshCredentials(OAuthCredentialsProvider.java:133)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.loadCredentials(OAuthCredentialsProvider.java:121)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.applyCredentials(OAuthCredentialsProvider.java:79)
```

</p>
</details>

**Environment:**
- OS: MacOS, Linux
- Zeebe Version: from main branch
- Configuration: `OAuthCredentialsProvider`

### PR DESCRIPTION
**Pull request by aivinog1**
## Description

I introduced some locking mechanisms in the OAuth2 integration in Zeebe Client Java to make it thread-safe.
I don't like the current solution but I don't have nice ideas on how to fix it in another way.

In the future, we could allow users to move [this call](https://github.com/camunda/zeebe/blob/main/clients/java/src/main/java/io/camunda/zeebe/client/impl/ZeebeCallCredentials.java#L49) inside their `CredentialsProvider`, because in the `io.grpc.CallCredentials` class [JavaDoc says](https://github.com/grpc/grpc-java/blob/5be17e8b22d1d3eb99ea92140af2297734100e63/api/src/main/java/io/grpc/CallCredentials.java#L39-L54): 
> If metadata is not immediately available, e.g., needs to be fetched from the network, the implementation may give the applier to an asynchronous task which will eventually call the applier. The RPC proceeds only after the applier is called.

And we could build a nice asynchronous pipeline without locking (in my mind it could be based on a queue or something like this: we just push some actions in the queue and a single consumer of this queue does all logic (thus these consumers calls could be lock-free and not thread-safe, but the whole construction will be thread safe)).

## Related issues

closes #11885 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java [hunk 1]
```diff
@@ -108,17 +121,22 @@ public boolean shouldRetryRequest(final Throwable throwable) {
   private void loadCredentials() throws IOException {
     Optional<ZeebeClientCredentials> cachedCredentials;
 
+    cacheLock.lock();
```
Comment (remcowesterhoud) [inline clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java:124]: ❓ Do we need this lock if we've already made the cache thread-safe?

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks for the fix @aivinog1! I have a question about your use of locks. There also is an issue with the CI. Would you mind having a look at this as well?

File: clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java (hunk 1 above)
Comment (aivinog1) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java:124]: Hey @remcowesterhoud!

Yes, you made a good point. Let me rethink it a little bit.

Comment (aivinog1) [reply clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java]: @remcowesterhoud Can you please take another look? I refactored it a bit.

**Review by remcowesterhoud — APPROVED** [review]
Sorry for taking so long to review this again. Love the solution you went for 👍 Thanks @aivinog1 !

bors merge

**Comment by aivinog1** [comment]
> * Test summary

@remcowesterhoud Could you restart the build, please ? :)

**Comment by remcowesterhoud** [comment]
👍 

bors retry

**Pull request merged**